### ISSUE TITLE
Implement a Pattern for Tracking Versions of content

### ISSUE THREAD

**Issue by flyingzumwalt** [ipfs/in-web-browsers#10]
(probably out of scope)

prerequisite: #27

## Background 

@gozala

> Some of the ideas we (me and Patryk) being exploring seem to assume that there is a way to see an every version of the IPFS content. In other words it would be nice to have a changelog for IPNS up2017-02-01s so it’s not just here is what the current version is, but also here all the previous version that existed. I remember mention of the commit objects in white paper so I assume there is a way to up2017-02-01 IPNS pointer with a commit object, but I can’t really figure out how or if I’m actually getting it right. I think something along the lines of http://docs.datproject.org/sleep is what I’m looking for.

@jbenet: 

>  Yes, we have given this a lot of thought, and are returning to it this and next quarter. it's not easy to get this right because what we choose can block many applications. Meaning that "one fully-contained versioning strategy" works for 20% of use cases we've looked at at most. One clear example is that data-center applications that expect to mutate names on the order of <1ms will want something that works a bit differently than apps that require much stronger security (eg censorship resistance that requires timestamp to the bitcoin blockchain, DNS, or some equivalent level of security) but can tolerate only changing names on the order of <100s (like most DNS names).
> 
> This actually decomposes to two different problems:
> 
> - How apps want to do versioning (security, consistency, and dev UX, implications):
>   - Commit graphs (like git)
>   - Commutable patches (like darcs)
>   - CRDTs (riak, orbit, google docs, google internals, the future)
>   - consensus (blockchains, etc).
> 
> - How apps want to do naming (security, consistency, dev UX, and ownership implications):
>   - slow public key (ipns, sfs) with consensus (strong consistency, >10s up2017-02-01s, available only in some consensus model)
>   - fast public key (ipns, sfs) without consensus (weaker consistency, >1us up2017-02-01s, available disconnected networks -- dhts, pubsub, etc)
>   - DNS naming (strong consistency, >60s up2017-02-01s)
>   - blockchain naming (ENS, blockstack, etc).
> 
> In our research, lots of apps DO NOT want to manage their versions manually, want convergent replication, and should be using CRDTs and things like orbit-db. Some subset DO want direct control over versioning and want commit graphs (like git, dat, etc), so for those we will expose direct versioning logs that can be indexed in a couple of good ways. (eg binomial heaps, etc)
> 
> BTW, i think the big hump that we need to communicate better is the transition from "apps store data in files" to "apps store data directly, can build files out of data", and the IPFS name isn't helping a ton here.

**Comment by lgierth** [comment]
Continued in #27
